### PR TITLE
Use project's java runtime to launch the application

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

This PR requires https://github.com/microsoft/vscode-java-debug/pull/758

### CODE REVIEW

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveJavaExecutableHandler.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * Resolve the java executable path from the project's java runtime.
+     */
+    public static String resolveJavaExecutable(List<Object> arguments) throws Exception {
+        try {
+            String mainClass = (String) arguments.get(0);
+            String projectName = (String) arguments.get(1);
+            IJavaProject targetProject = null;
+            if (StringUtils.isNotBlank(projectName)) {
+                targetProject = JdtUtils.getJavaProject(projectName);
+            } else {
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveJavaExecutableHandler.java]: If a project is not specified, should we just use whatever JVM that the debugger is using to launch the app?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 2]
```diff
@@ -108,6 +177,7 @@ public static IDebugSession launch(VirtualMachineManager vmManager,
      *             with an error before a connection could be established.
      */
     public static IDebugSession launch(VirtualMachineManager vmManager,
+            String javaExec,
```
[the remark this answers is not in the document]
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: Nice catch!

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveJavaExecutableHandler.java (hunk 1 above)
Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveJavaExecutableHandler.java]: In launch.json, `project` info is optional. In the case of undefined, the debugger will try to resolve the project info from the main class. See the `else` branch.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 3]
```diff
@@ -164,7 +234,12 @@ public static IDebugSession launch(VirtualMachineManager vmManager,
             arguments.get(ENV).setValue(encodeArrayArgument(envVars));
         }
 
-        if (StringUtils.isNotEmpty(DebugSettings.getCurrent().javaHome)) {
+        if (isValidJavaExec(javaExec)) {
+            String vmExec = new File(javaExec).getName();
+            String javaHome = new File(javaExec).getParentFile().getParentFile().getAbsolutePath();
```
[the remark this answers is not in the document]
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: This is in order to keep compatible with JDI Socket LaunchingConnectorImpl https://github.com/eclipse/eclipse.jdt.debug/blob/master/org.eclipse.jdt.debug/jdi/org/eclipse/jdi/internal/connect/SocketLaunchingConnectorImpl.java#L188, where it uses `javaHome + separator + vmExec` to construct the command line.

Currently when launching in debug console, we're still using SocketLaunchingConnector to construct the command line, so need resolve the javaHome and vmExec from the path and pass them as the connector arguments.

Another option is to use ListeningConnector instead, like what launchInTerminal did, the debugger can leverage the javaExec path directly. Let's track the refactor improvement in the issue https://github.com/microsoft/vscode-java-debug/issues/760

**Comment by testforstephen** [comment]
@akaroml PR updated

**Pull request merged**